core/devmode/src/main/java/io/quarkus/dev/RuntimeUpdatesProcessor.java

```diff
@@ -112,31 +117,38 @@ public void addPreScanStep(Runnable runnable) {
 
     boolean checkForChangedClasses() throws IOException {
         boolean hasChanges = false;
-
-        for (DevModeContext.ModuleInfo i : context.getModules()) {
-            for (String sourcePath : i.getSourcePaths()) {
+        for (DevModeContext.ModuleInfo module : context.getModules()) {
+            final List<Path> moduleChangedSourceFilePaths = new ArrayList<>();
+            for (String sourcePath : module.getSourcePaths()) {
                 final Set<File> changedSourceFiles;
                 try (final Stream<Path> sourcesStream = Files.walk(Paths.get(sourcePath))) {
                     changedSourceFiles = sourcesStream
                             .parallel()
-                            .filter(p -> matchingHandledExtension(p).isPresent())
-                            .filter(p -> wasRecentlyModified(p))
+                            .filter(p -> matchingHandledExtension(p).isPresent() && wasRecentlyModified(p))
                             .map(Path::toFile)
                             //Needing a concurrent Set, not many standard options:
                             .collect(Collectors.toCollection(ConcurrentSkipListSet::new));
                 }
                 if (!changedSourceFiles.isEmpty()) {
                     log.info("Changed source files detected, recompiling " + changedSourceFiles);
                     try {
+                        final Set<Path> changedPaths = changedSourceFiles.parallelStream()
```

Comment:
This should not be using parallelStream(). parallelStream is a win if you have thousands of items, for small collections it is not worth it, especially when just calling .toPath